contents/javascript/staticTypingWithTypescript.md

```diff
@@ -0,0 +1,54 @@
+<frontmatter>
+  title: Static Typing with TypeScript
+  header: pagetop.md
+  footer: footer.md
+  head: head.md
+  siteNav: mainNav.md
+  pageNav: 3
+</frontmatter>
+
+<div class="website-content">
+
+{{ booktitle | safe }}
+
+# Static Typing with TypeScript
+
+**Author(s): [Lu Yang Kenneth](https://github.com/luyangkenneth)**
+
+Reviewer(s): WIP FILL THIS OUT LATER
+
+
+## What is Static Typing?
+
+**Statically typed languages** like Java, Go, and C++ are able to catch type-related errors at compile time, while in **dynamically typed languages** like JavaScript, such errors are not as easily discoverable because the types of variables are only known at runtime.
+
+To help identify these problems without having to run any code, many developers have adopted technologies like Flow and TypeScript, which enable static type checking for codebases written in JavaScript.
+
+Equivalent type checking tools exist for other dynamically typed languages like Python and Ruby as well, which suggests that this issue is not specific to the JavaScript community. The decision to move towards a statically typed codebase is usually motivated by the following benefits:
+
+- Static type checking **reduces type-related errors during the development process**. These errors can be expensive to fix if they leak into production and cause downtime in your application, especially for teams that value stability over short-term productivity. A type checker provides an automatic way to verify the type safety and correctness of your application _before_ it goes live.
+
+- Static type checking can **make developers more productive**. When paired with the appropriate IDE support, it enables features such as auto-completion and offers instant feedback by incrementally rechecking your code as you type.
+
+- Static type checking **improves collaboration on a large codebase**. Explicitly declared types allow developers to refactor with greater confidence, knowing that API boundaries are enforced by the type checker. If you change a method signature or the shape of an object, you can quickly discover the other parts of the codebase that also need to be changed, as they would be surfaced as errors by the type checker.
+
+
+### What's the difference between Flow and TypeScript?
```

Comment:
I'm not sure if discussing the difference between Flow and TypeScript is appropriate here, especially since there is no context as to what Flow is yet. It may be confusing for someone new to this subject to introduce it so early. Perhaps you could consider placing this at the end instead as a section on alternatives to TypeScript?

Reply:
> especially since there is no context as to what Flow is yet

I did mention it in a paragraph on top: `...technologies like Flow and TypeScript, which enable static type checking for codebases written in JavaScript.`

Hmm I wanted to use this an opportunity to state what TypeScript is upfront (a programming language; not quite the same as a type checking tool) in the same breath as Flow so that there's a super clear distinction about their high-level differences. And then use this also as a segue to `As the industry seems to be shifting towards TypeScript as the top choice, in this article I will introduce how TypeScript works and how you can get started with TypeScript.`

Placing this at the end could of course still work, but that would mean duplicating `TypeScript is a statically typed superset of JavaScript that compiles to plain JavaScript.` again after the article has already gone through TypeScript Basics.

---

Maybe what I'll do is to restructure this to:
- What is Static Typing? (no mention of Flow / TypeScript; all discussion here is generic)
- Adding Static Typing to JavaScript ("There are two popular ways to do this! Here are their differences! Now I will talk more about TypeScript!")

^ Does this sound better to you?

Reply:
Yes, I think that makes it a little clearer why Flow is being introduced. However, I'm still not sure how valuable introducing Flow may be - you might also just add a note on it in the further reading section. But if you think it's useful, I think the structure you proposed makes sense.